### PR TITLE
feat: initial bookmarks UI

### LINKED ISSUE #22: Implement bookmarks
Bookmarks should act on the current tab, bookmarks should include:

- Folders (events, gurpurabs, themes, etc.)
- Dynamic bookmarks
- Regular shabad/dictionary bookmarks (user definable)
- Compilations / banis (e.g. nitnem, asa ki vaar)

Mockup example:
image

### LINKED ISSUE #124: Implement read-only bookmarks
These are bookmarks that are preexisting in the database.

Parent issue #22.

### Acceptance Criteria
- Component takes a list of bookmarks to render as input
- Screen appears from bottom or side, not as modal
- Should support rendering a list of "folders"
- Clicking on an item fires a callback with what was clicked on
- Tests

### Bonus
- Virtualisation using flatlists or similar for items

@bhajneet's todo:
- Share relevant mocks
- Create `collections.json` which is a list of all folders containing bani ids:
  - `[ { name: 'Nitnem', bookmarks: [ 1, 2, 3, 4 ] } ]`

### PR DESCRIPTION
**Pull request by AkalUstat**
### Summary of PR
try to implement bookmarks (again) with new design

### Mocks
<details>
<summary><b>iOS</b></summary>
iOS mockup
</details>

<details>
<summary><b> Android </b></summary>
Android mockup
</details>

### Time spent on PR
2 hours: @AkalUstat 
6 hours: @saihaj 

### Linked issues
Fix #22, fix #124

### Reviewers
@Harjot1Singh 

### Todo
- [x] fix TS errors
- [x] style component fine tuning will be done later. #129 brings many styles with it so we can do once that is merged.
- [x] Write maybe more tests for Bookmarks @Harjot1Singh help needed
~[ ] header~ once #129 is merged we can take pieces from there to help

### CODE REVIEW

File: src/components/BookmarksList.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  onPress: () => void,
+}
+
+const BookmarkListItem = ( {
+  item,
+  isFolder,
+  onPress,
+}: BookmarkListItemProps ) => (
+  <Pressable style={styles.BookmarkContainer} onPress={onPress}>
+    <Text>{item?.name ?? item}</Text>
```
Comment (AkalUstat) [inline src/components/BookmarksList.tsx:30]: issue with `.name` here

File: src/components/BookmarksList.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+const BookmarksList = ( { data, onItemPress }: BookmarksListProps ) => (
+  <View>
+    {data.map( ( item ) => {
+      const isFolder = checkIsFolder( item )
+      return (
+        <BookmarkListItem
+          item={item}
+          isFolder={isFolder}
+          onPress={() => onItemPress( isFolder, isFolder ? item?.name : item )}
```
Comment (AkalUstat) [inline src/components/BookmarksList.tsx:47]: same issue here

File: src/screens/BookmarksScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  const route = useRoute<Route>()
+  const navigation = useNavigation<Navigation>()
+
+  const { folderData }: {folderData: ( Folder | FolderItem )[]} = route.params
+
+  const onItemPress = ( isFolder: boolean, name: string ) => {
+    if ( isFolder ) {
+      navigation.push( Screens.Bookmarks,
+        {
+          folderData: folderData.find( ( folder ) => folder?.name === name )?.bookmarks,
```
Comment (AkalUstat) [inline src/screens/BookmarksScreen.tsx:23]: same issue here

File: src/components/BookmarksList.tsx (hunk 1 above)
Comment (Harjot1Singh) [reply src/components/BookmarksList.tsx:30]: Issue comes from effectively two types spawning from same expression, string or object.

I'd also actually just structure bookmarks like this:

```
type NavigationItem = {
  name: string,
  destinations?: NavigationItem[],
}
```

and so you know if `navigationItem.destinations` is undefined, you've reached the end of drilling into folders and it's a list item.

This approach also is for writing a generic structure for navigation.

**Comment by saihaj** [comment]
@Harjot1Singh can you please take a look here. Little confused. Props are kinda passed down but not working as expected. 
Kapture 2021-01-13 at 21 10 50

**Comment by saihaj** [comment]
@bhajneet
Kapture 2021-07-31 at 21 21 38

File: src/components/Bookmarks/List.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import React from 'react'
+import { render, fireEvent } from '@testing-library/react-native'
+
+import Wrapper, { WrapperProps } from '../../lib/NavigatorContext'
+import defaultFolderData from '../../defaults/collections.json'
+import Screens from '../../lib/screens'
+
+import BookmarksList from './List'
+
+const bookmarkScreen = ( { children }:WrapperProps ) => (
```
Comment (Harjot1Singh) [inline src/components/Bookmarks/List.spec.tsx:10]:
```suggestion
const bookmarkScreen = ( { children }: WrapperProps ) => (
```

unfortunately I'm not sure there's a lint rule for TS spacing

File: src/components/Bookmarks/List.tsx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        {
+          folderData:
+          folderData.find( ( folder ) => folder.name === name )?.bookmarks
+          || folderData,
+          currentFolder: name,
+        } )
+    } else {
+      Alert.alert( `you clicked on ${name}` )
+    }
+  }
```
Comment (Harjot1Singh) [inline src/components/Bookmarks/List.tsx:41]: To reduce the level of indentation, this pattern is better:

```
if (!isFolder) {
  Alert.alert('')
  return
}

// do rest
```

**Review by Harjot1Singh — CHANGES_REQUESTED** [review]

File: src/components/Bookmarks/List.spec.tsx (hunk 4 above)
Comment (saihaj) [reply src/components/Bookmarks/List.spec.tsx:10]: unfortunately not Eslint or prettier don't pick it up.

**Review by Harjot1Singh — APPROVED** [review]

**Pull request merged**